gobblin-config-management/gobblin-config-core/src/main/java/gobblin/config/common/impl/InMemoryTopology.java

```diff
@@ -0,0 +1,140 @@
+package gobblin.config.common.impl;
+
+import gobblin.config.store.api.ConfigKeyPath;
+
+import java.util.Collection;
+import java.util.List;
+
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.ListMultimap;
+
+/**
+ * InMemoryTopology will return stale data if the internal config store is Not {@ConfigStoreWithStableVersioning}
+ * 
+ * @author mitu
+ *
+ */
+public class InMemoryTopology implements ConfigStoreTopologyIntf {
+
+  private final ConfigStoreBackedTopology fallback;
+
+  private final ListMultimap<ConfigKeyPath, ConfigKeyPath> childrenMap = ArrayListMultimap.create();
+  private final ListMultimap<ConfigKeyPath, ConfigKeyPath> ownImportMap = ArrayListMultimap.create();
+  private final HashMultimap<ConfigKeyPath, ConfigKeyPath> ownImportedByMap = HashMultimap.create();
+  private final ListMultimap<ConfigKeyPath, ConfigKeyPath> recursiveImportMap = ArrayListMultimap.create();
+  private final HashMultimap<ConfigKeyPath, ConfigKeyPath> recursiveImportedByMap = HashMultimap.create();
+
+  public InMemoryTopology(ConfigStoreBackedTopology fallback) {
+    this.fallback = fallback;
+  }
+
+  private void loadRawTopologyFromConfigStore() {
+    // breath first search the whole topology
+    // calls to retrieve cache / set cache if not present
+    Collection<ConfigKeyPath> nodeQueue = this.getChildren(SingleLinkedListConfigKeyPath.ROOT);
+
+    while(!nodeQueue.isEmpty()){
+      ConfigKeyPath configKeyPath = nodeQueue.iterator().next();
+      nodeQueue.remove(configKeyPath);
+
+      // calls to retrieve cache / set cache if not present
+      List<ConfigKeyPath> ownImports = this.getOwnImports(configKeyPath);
+      for(ConfigKeyPath p: ownImports){
+        this.ownImportedByMap.put(p, configKeyPath);
+      }
+
+      List<ConfigKeyPath> recursiveImports = null;
+      // result may in the cache already 
+      if(this.recursiveImportMap.containsKey(configKeyPath)){
+        recursiveImports = this.recursiveImportMap.get(configKeyPath);
+      }
+      else {
+        recursiveImports = SimpleImportChainResolver.getImportsRecursively(this.fallback.getConfigStore(), 
```

Comment:
This is very inefficient. If you have the import chain A -> B -> C -> D, you'll compute the recursive imports for B twice (once for A and once for B itself), C -- three times, etc.

Further, SimpleImportChainResolver.getImportsRecursively does not utilize the ownImports cache.

I suggest you do a two-pass algorithm. In the first pass, you scan the children hierarchy and fill-in the ownImports/ownImportedBy caches. In the second pass, you iterate over the keys of ownImports to compute the recursive imports. You should also memoization so you calculate the recursive imports for a given key path only once.